### PR TITLE
Several fixes and cleanup

### PR DESCRIPTION
**Pull request by pimlie**
Hi, actually I only needed to have the debug statements to be printed without htmlentities as I am connecting with the webservice through a console command. But while I was at it I fixed some other things as well, see separate commit messages.

Also fixed / added some tests. We could probably add quite a bit more though...

(run `composer test` to quickly test or `composer cs` to quickly check code styles)

### CODE REVIEW

**Comment by Silentscripter** [comment]
Hi @pimlie thank's a lot for your contribution, your work has been impressive :) I'll try to review the code in the next two days.

**Comment by pimlie** [comment]
No problem, thanks for your initial work.

In my last commit I also added a separate exception for requests that includes the prestashop xml response (if any). The prestashop error message can be quite usefull. 
Also added the status code to the exception codes, this way you can quickly do `if edit == 404 then add`.

File: src/PrestashopWebServiceLibrary.php [hunk 1]
```diff
@@ -44,155 +50,232 @@ class PrestashopWebServiceLibrary
...
+        if (isset($messages[$request['status_code']])) {
+            $xml = null;
+            if ($request['response']) {
+                $xml = $this->parseXML($request['response'], true);
+            }
+
+            throw new PrestashopWebServiceRequestException($messages[$request['status_code']], $request['status_code'], $xml);
+        } else {
+            $exception = 'This call to PrestaShop Web Services returned an unexpected HTTP status of: ';
+            $exception.= $status_code;
```
Comment (Silentscripter) [inline src/PrestashopWebServiceLibrary.php]: `$status_code` is not defined, maybe you meant `$request[status_code]` ?

**Review by Silentscripter — CHANGES_REQUESTED** [review]
I believe there's an error in calling the status code as highlighted in the comment, could you check please?

**Comment by pimlie** [comment]
@Silentscripter You are correct, thanks for spotting that. I added a test to make sure it doesnt happen again ;)

**Review by Silentscripter — APPROVED** [review]

**Comment by Silentscripter** [comment]
There's something I'd like to discuss with you before merging:
1. I get an error during tests at `it_can_perform_a_get_request()` test with this error:
```
There was 1 error:

1) Protechstudio\PrestashopWebService\Tests\PrestashopWebServiceTest::it_can_perform_a_get_request
Protechstudio\PrestashopWebService\Exceptions\PrestashopWebServiceException: HTTP XML response is not parsable: array (
  0 => 
  LibXMLError::__set_state(array(
     'level' => 3,
     'code' => 4,
     'column' => 1,
     'message' => 'Start tag expected, \'<\' not found
',
     'file' => '',
     'line' => 1,
  )),
)

/home/vagrant/code/laravel-prestashop-webservice/src/PrestashopWebServiceLibrary.php:268
/home/vagrant/code/laravel-prestashop-webservice/src/PrestashopWebServiceLibrary.php:370
/home/vagrant/code/laravel-prestashop-webservice/tests/PrestashopWebServiceTest.php:25
```
Is it related to me missing something in test configuration or is there an error in the response mock file?

2. You renamed `PrestaShopWebServiceException` to `PrestashopWebServiceException`, although I would agree with the change itself this must be considered a breaking change since it will break try / catch blocks. The problem is I'm not that sure this change is enough to justifiy a new major version. What do you think about this? Shouldn't we move this change in another patch when adding more breaking changes maybe?

**Comment by pimlie** [comment]
1. Could this be because `short_open_tags` is enabled for php? Then the `<?` in `<?xml` will be treated as a php opening tag.

2. I wondered about that too, but it appears that class names are case insensitive in php. I tested it and for me it doesnt give any error whether I use `Prestashop` or `PrestaShop`.

File: tests/PrestashopWebServiceTest.php [hunk 2]
```diff
@@ -1,18 +1,138 @@
...
     public function it_is_correctly_installed()
     {
         $this->assertInstanceOf(PrestashopWebService::class, Prestashop::getFacadeRoot());
     }
+    
+    /** @test */
+    public function it_can_perform_a_get_request()
+    {
+        $requestResponseStub = require('requests/category-schema.php');
+        $ps = $this->getMockedLibrary('executeCurl', $requestResponseStub);
```
Comment (Silentscripter) [inline tests/PrestashopWebServiceTest.php]: `$ps = $this->getMockedLibrary('executeCurl', $requestResponseStub);` is correct? Shouldn't it be `$ps = $this->getMockedLibrary('executeRequest', $requestResponseStub);`

**Review by Silentscripter — COMMENTED** [review]
Could you verify if this is correct?

File: tests/PrestashopWebServiceTest.php (hunk 2 above)
Comment (Silentscripter) [reply tests/PrestashopWebServiceTest.php]: 1. Changing it to `executeRequest` the error changes:
```
There was 1 error:

1) Protechstudio\PrestashopWebService\Tests\PrestashopWebServiceTest::it_can_perform_a_get_request
ErrorException: Undefined index: status_code

/home/vagrant/code/laravel-prestashop-webservice/src/PrestashopWebServiceLibrary.php:78
/home/vagrant/code/laravel-prestashop-webservice/src/PrestashopWebServiceLibrary.php:369
/home/vagrant/code/laravel-prestashop-webservice/tests/PrestashopWebServiceTest.php:25

ERRORS!
Tests: 8, Assertions: 14, Errors: 1.
```
2. As far as I know this depends from system as Windows is case-insensitive while linux is case-sensitive but I'll conduct a test to see if I'm correct.

Comment (pimlie) [reply tests/PrestashopWebServiceTest.php]: No, mocking `executeCurl` is correct in this test. Part of the test is whether `executeRequest` handles a curl response correctly. Thats the reason I separate the curl logic to a separate method, this way we can easily test/mock the curl stuff to easily test our logic that handles that curl response. 
What you can to check if the short open tag is the problem is to add `ini_set('short_open_tags', 0);` in the test method. Or change `<?xml` to `'<' . '?xml'.

If you really want to be sure, you could maybe add travis (linux) / appveyor (windows) configuration for automatic build tests for php5.4, php7.0. Its the easiest to check things across different systems. Just copy and adapt the config from other projects :)

Comment (Silentscripter) [reply tests/PrestashopWebServiceTest.php]: I'm actually testing on Vagrant homestead with php7.1 and checked that short_open_tags is off as it should be so it's not the problem. There's something that is strange anyway, if you read the error (first one) it declares `'file' => '',`.
I was already thinking about adding CI with travis but didn't have time yet. I'll take a look now

Comment (pimlie) [reply tests/PrestashopWebServiceTest.php]: Could it maybe be an issue with linebreaks, so `\n` vs `\r\n`? See the new test I added

**Comment by Silentscripter** [comment]
The new test triggers another error, it seems like there are some problems capturing the `category-schema.php` content:
```
There were 2 errors:

1) Protechstudio\PrestashopWebService\Tests\PrestashopWebServiceTest::test_request_is_correct
ErrorException: Undefined offset: 1

/home/vagrant/code/laravel-prestashop-webservice/tests/PrestashopWebServiceTest.php:24

2) Protechstudio\PrestashopWebService\Tests\PrestashopWebServiceTest::it_can_perform_a_get_request
Protechstudio\PrestashopWebService\Exceptions\PrestashopWebServiceException: HTTP XML response is not parsable: array (
  0 => 
  LibXMLError::__set_state(array(
     'level' => 3,
     'code' => 4,
     'column' => 1,
     'message' => 'Start tag expected, \'<\' not found
',
     'file' => '',
     'line' => 1,
  )),
)

/home/vagrant/code/laravel-prestashop-webservice/src/PrestashopWebServiceLibrary.php:268
/home/vagrant/code/laravel-prestashop-webservice/src/PrestashopWebServiceLibrary.php:370
/home/vagrant/code/laravel-prestashop-webservice/tests/PrestashopWebServiceTest.php:36

ERRORS!
Tests: 9, Assertions: 14, Errors: 2.
```
Anyway I added CI with travis so if you merge and push a new commit we may check this build and see how it performs on Travis.

**Comment by pimlie** [comment]
Travis doesnt work as its missing composer dependencies, please try to add the following to the travis config:
```
php:
  - 7.2
sudo: false

before_script:
  - travis_retry composer self-update
  - travis_retry composer install --no-interaction

script: vendor/bin/phpunit
```

**Comment by Silentscripter** [comment]
Hi @pimlie , sorry for the delay, kinda busy these weeks. I've pushed the fix for travis as suggested. Would you please update your branch and verify? Thanks.

**Comment by pimlie** [comment]
According to [this](https://stackoverflow.com/questions/17606874/trigger-a-travis-ci-rebuild-without-pushing-a-commit/17624403#17624403) SO, closing and opening the PR should also trigger Travis.

**Comment by pimlie** [comment]
And lets see if it works
--edit--
Yes, it works :)

**Comment by pimlie** [comment]
Hmm, the orchestra testbench suite doesnt support php 5.6 it seems

**Comment by Silentscripter** [comment]
Hi @pimlie , would you evaluate to push a commit to restore `PrestashopWebServiceException` to `PrestaShopWebServiceException` ? This way I can tag the update with a minor patch version instead of a major one. We may plan a new milestone with new functionalities (I've some ideas) where to include this breaking change.

**Comment by pimlie** [comment]
Sure, this commit should fix that. For the next major release you can then just remove `src/PrestaShopWebserviceException.php` and have `src/Exceptions/PrestashopWebServiceException.php` extend `\Exception` instead.

**Pull request merged**